test/state/precompiles.cpp

```diff
@@ -292,6 +301,40 @@ ExecutionResult blake2bf_execute(const uint8_t* input, [[maybe_unused]] size_t i
     return {EVMC_SUCCESS, sizeof(h)};
 }
 
+ExecutionResult bls12_g1add_execute(
+    const uint8_t* input, size_t input_size, uint8_t* output, size_t output_size) noexcept
+{
+    static_assert(std::endian::native == std::endian::little,
+        "bls12_g1add only works correctly on little-endian architectures");
+
+    if (input_size != 256)
+        return {EVMC_PRECOMPILE_FAILURE, 0};
+
+    if (output_size != 128)
+        return {EVMC_PRECOMPILE_FAILURE, 0};
+
+    memset(output, 0, output_size);
+
+    constexpr auto verify_element = [](const uint8_t* p) {
+        return intx::be::unsafe::load<intx::uint512>(p) < crypto::bls::FieldPrime;
+    };
+
+    if (!verify_element(input))
+        return {EVMC_PRECOMPILE_FAILURE, 0};
+    if (!verify_element(&input[64]))
+        return {EVMC_PRECOMPILE_FAILURE, 0};
+    if (!verify_element(&input[128]))
+        return {EVMC_PRECOMPILE_FAILURE, 0};
+    if (!verify_element(&input[192]))
+        return {EVMC_PRECOMPILE_FAILURE, 0};
+
+    if (!crypto::bls::g1_add(&output[16], &output[16 + 64], &input[16], &input[16 + 64],
```

Comment:
I think I would just send 64 byte field element down the `crypto` function and validate there as `uint512`. Or even a point as 128 bytes... Otherwise we have validation split into two parts.

Reply:
Implemented in https://github.com/ethereum/evmone/pull/984